Source/Classes/AnimatedImages/PINCachedAnimatedImage.m

```diff
@@ -0,0 +1,426 @@
+            if (beginningKeepRange.location != NSNotFound) {
+                shouldKeepFrame |= NSLocationInRange(idx, beginningKeepRange);
+            }
+            if (shouldKeepFrame == NO) {
+                [removedFrames addIndex:idx];
+                [self->_frameCache removeObjectForKey:@(idx)];
+                PINLog(@"Removing: %lu", (unsigned long)idx);
+            }
+        }];
+        [_cachedOrCachingFrames removeIndexes:removedFrames];
+    }];
+}
+
+- (void)l_cacheFrame:(NSUInteger)frameIndex
+{
+    if ([_cachedOrCachingFrames containsIndex:frameIndex] == NO) {
+        PINLog(@"Requesting: %lu", (unsigned long)frameIndex);
+        [_cachedOrCachingFrames addIndex:frameIndex];
+        _playbackReady++;
+        
+        dispatch_async(_cachingQueue, ^{
+            CGImageRef imageRef = [self->_animatedImage imageAtIndex:frameIndex cacheProvider:self];
+            PINLog(@"Generating: %lu", (unsigned long)frameIndex);
+
+            if (imageRef) {
+                [self->_lock lockWithBlock:^{
+                    [self->_frameCache setObject:(__bridge id _Nonnull)(imageRef) forKey:@(frameIndex)];
+                    self->_playbackReady--;
+                    NSAssert(self->_playbackReady >= 0, @"playback ready is less than zero, something is wrong :(");
+                    
+                    PINLog(@"Frames left: %ld", (long)_playbackReady);
+                    
+                    dispatch_block_t notify = nil;
+                    if (self->_playbackReady == 0 && self->_notifyOnReady) {
+                        self->_notifyOnReady = NO;
+                        if (self->_playbackReadyCallback) {
+                            notify = self->_playbackReadyCallback;
+                            [_operationQueue addOperation:^{
+                                notify();
+                            }];
+                        }
+                    }
+                }];
+            }
+        });
+    }
+}
+
+// Returns the number of frames that should be cached
+- (NSUInteger)framesToCache
+{
+    unsigned long long totalBytes = [NSProcessInfo processInfo].physicalMemory;
+    NSUInteger framesToCache = 0;
+    
+    NSUInteger frameCost = _animatedImage.bytesPerFrame;
+    if (frameCost * _animatedImage.frameCount < totalBytes / 250) {
+        // If the total number of bytes takes up less than a 250th of total memory, lets just cache 'em all.
+        framesToCache = _animatedImage.frameCount;
+    } else if (frameCost < totalBytes / 1000) {
+        // If the cost of a frame is less than 1000th of physical memory, cache 4 frames to smooth animation.
+        framesToCache = kFramesToRenderForLargeFrames;
+    } else if (frameCost < totalBytes / 500) {
+        // Oooph, lets just try to get ahead of things by one.
+        framesToCache = kFramesToRenderMinimum;
+    } else {
+        // No caching :(
+        framesToCache = 0;
+    }
```

Comment:
Good comments. Might be good in the future to add awareness of multiple live instances of `PINCachedAnimatedImage` in the future. If you have many live instances in the first bucket above we could still drive the system to run out of memory. You could have each live instance get a memory quota that grows/shrinks with the number of live instances.